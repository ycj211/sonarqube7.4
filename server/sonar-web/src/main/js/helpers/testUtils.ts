/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow, ShallowRendererProps, ShallowWrapper, ReactWrapper } from 'enzyme';
import { InjectedRouter } from 'react-router';
import { IntlProvider } from 'react-intl';

export const mockEvent = {
  target: { blur() {} },
  currentTarget: { blur() {} },
  preventDefault() {},
  stopPropagation() {}
};

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', { ...mockEvent, ...event });
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export function keydown(keyCode: number): void {
  const event = new KeyboardEvent('keydown', { keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.body.clientWidth/clientHeight` are getters by default, so we need to redefine them
  // pass `configurable: true` to allow to redefine the properties multiple times
  if (width) {
    Object.defineProperty(document.body, 'clientWidth', { configurable: true, value: width });
  }
  if (height) {
    Object.defineProperty(document.body, 'clientHeight', { configurable: true, value: height });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

// Create the IntlProvider to retrieve context for wrapping around.
const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();
export function shallowWithIntl<C extends React.Component>(
  node: React.ReactElement<any>,
  options: ShallowRendererProps = {}
) {
  return shallow<C>(node, { ...options, context: { intl, ...options.context } });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}

export function mockRouter(overrides: { push?: Function; replace?: Function } = {}) {
  return {
    createHref: jest.fn(),
    createPath: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    isActive: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    setRouteLeaveHook: jest.fn(),
    ...overrides
  } as InjectedRouter;
}
