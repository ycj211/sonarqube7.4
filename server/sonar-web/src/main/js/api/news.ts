/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getCorsJSON } from '../helpers/request';

interface PrismicRef {
  id: string;
  ref: string;
}

export interface PrismicNews {
  data: { title: string };
  last_publication_date: string;
  uid: string;
}

const PRISMIC_API_URL = 'https://sonarsource.cdn.prismic.io/api/v2';

export function fetchPrismicRefs() {
  return getCorsJSON(PRISMIC_API_URL).then((response: { refs: Array<PrismicRef> }) => {
    const master = response && response.refs.find(ref => ref.id === 'master');
    if (!master) {
      return Promise.reject('No master ref found');
    }
    return Promise.resolve(master);
  });
}

export function fetchPrismicNews(data: {
  accessToken: string;
  ps?: number;
  ref: string;
  tag?: string;
}) {
  const q = ['[[at(document.type, "blog_sonarsource_post")]]'];
  if (data.tag) {
    q.push(`[[at(document.tags,["${data.tag}"])]]`);
  }
  return getCorsJSON(PRISMIC_API_URL + '/documents/search', {
    access_token: data.accessToken,
    orderings: '[document.first_publication_date desc]',
    pageSize: data.ps || 1,
    q,
    ref: data.ref
  }).then(({ results }: { results: Array<PrismicNews> }) => results);
}
