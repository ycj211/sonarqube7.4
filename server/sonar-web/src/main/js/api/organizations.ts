/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, postJSON } from '../helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import {
  Organization,
  OrganizationBase,
  Paging,
  OrganizationMember,
  Extension
} from '../app/types';

export function getOrganizations(data: {
  organizations?: string;
  member?: boolean;
}): Promise<{
  organizations: Organization[];
  paging: Paging;
}> {
  return getJSON('/api/organizations/search', data).catch(throwGlobalError);
}

export function getOrganization(key: string): Promise<Organization | undefined> {
  return getJSON('/api/organizations/search', { organizations: key }).then(
    r => r.organizations.find((o: Organization) => o.key === key),
    throwGlobalError
  );
}

interface GetOrganizationNavigation {
  adminPages: Extension[];
  canUpdateProjectsVisibilityToPrivate: boolean;
  isDefault: boolean;
  pages: Extension[];
}

export function getOrganizationNavigation(key: string): Promise<GetOrganizationNavigation> {
  return getJSON('/api/navigation/organization', { organization: key }).then(
    r => r.organization,
    throwGlobalError
  );
}

export function createOrganization(
  data: OrganizationBase & { installationId?: string }
): Promise<Organization> {
  return postJSON('/api/organizations/create', data).then(r => r.organization, throwGlobalError);
}

export function updateOrganization(key: string, changes: OrganizationBase) {
  return post('/api/organizations/update', { key, ...changes }).catch(throwGlobalError);
}

export function deleteOrganization(key: string) {
  return post('/api/organizations/delete', { key }).catch(throwGlobalError);
}

export function searchMembers(data: {
  organization?: string;
  p?: number;
  ps?: number;
  q?: string;
  selected?: string;
}): Promise<{ paging: Paging; users: OrganizationMember[] }> {
  return getJSON('/api/organizations/search_members', data).catch(throwGlobalError);
}

export function addMember(data: {
  login: string;
  organization: string;
}): Promise<OrganizationMember> {
  return postJSON('/api/organizations/add_member', data).then(r => r.user, throwGlobalError);
}

export function removeMember(data: { login: string; organization: string }) {
  return post('/api/organizations/remove_member', data).catch(throwGlobalError);
}

export interface OrganizationBilling {
  nclocCount: number;
  subscription: {
    plan?: {
      maxNcloc: number;
      price: number;
    };
    nextBillingDate?: string;
    status: 'active' | 'inactive' | 'suspended';
    trial: boolean;
  };
}

export function getOrganizationBilling(organization: string): Promise<OrganizationBilling> {
  return getJSON('/api/billing/show', { organization, p: 1, ps: 1 });
}
