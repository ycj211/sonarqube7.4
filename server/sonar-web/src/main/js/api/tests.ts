/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import throwGlobalError from '../app/utils/throwGlobalError';
import { Paging, TestCase, CoveredFile, BranchParameters } from '../app/types';
import { getJSON } from '../helpers/request';

export function getTests(
  parameters: {
    p?: number;
    ps?: number;
    sourceFileKey?: string;
    sourceFileLineNumber?: number;
    testFileKey: string;
    testId?: string;
  } & BranchParameters
): Promise<{ paging: Paging; tests: TestCase[] }> {
  return getJSON('/api/tests/list', parameters).catch(throwGlobalError);
}

export function getCoveredFiles(data: { testId: string }): Promise<CoveredFile[]> {
  return getJSON('/api/tests/covered_files', data).then(r => r.files, throwGlobalError);
}
