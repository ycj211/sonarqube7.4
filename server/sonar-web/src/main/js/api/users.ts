/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, postJSON, RequestData } from '../helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import { Paging, HomePage, CurrentUser, IdentityProvider, User } from '../app/types';

export function getCurrentUser(): Promise<CurrentUser> {
  return getJSON('/api/users/current');
}

export function changePassword(data: {
  login: string;
  password: string;
  previousPassword?: string;
}) {
  return post('/api/users/change_password', data).catch(throwGlobalError);
}

export interface UserGroup {
  default: boolean;
  description: string;
  id: number;
  name: string;
  selected: boolean;
}

export function getUserGroups(
  login: string,
  organization?: string,
  query?: string,
  selected?: string
): Promise<{ paging: Paging; groups: UserGroup[] }> {
  const data: RequestData = { login };
  if (organization) {
    data.organization = organization;
  }
  if (query) {
    data.q = query;
  }
  if (selected) {
    data.selected = selected;
  }
  return getJSON('/api/users/groups', data);
}

export function getIdentityProviders(): Promise<{ identityProviders: IdentityProvider[] }> {
  return getJSON('/api/users/identity_providers').catch(throwGlobalError);
}

export function searchUsers(data: {
  p?: number;
  ps?: number;
  q?: string;
}): Promise<{ paging: Paging; users: User[] }> {
  data.q = data.q || undefined;
  return getJSON('/api/users/search', data).catch(throwGlobalError);
}

export function createUser(data: {
  email?: string;
  local?: boolean;
  login: string;
  name: string;
  password?: string;
  scmAccount: string[];
}): Promise<void | Response> {
  return post('/api/users/create', data);
}

export function updateUser(data: {
  email?: string;
  login: string;
  name?: string;
  scmAccount: string[];
}): Promise<User> {
  return postJSON('/api/users/update', {
    ...data,
    scmAccount: data.scmAccount.length > 0 ? data.scmAccount : ''
  });
}

export function deactivateUser(data: { login: string }): Promise<User> {
  return postJSON('/api/users/deactivate', data).catch(throwGlobalError);
}

export function skipOnboarding(): Promise<void | Response> {
  return post('/api/users/skip_onboarding_tutorial').catch(throwGlobalError);
}

export function setHomePage(homepage: HomePage): Promise<void | Response> {
  return post('/api/users/set_homepage', homepage).catch(throwGlobalError);
}
