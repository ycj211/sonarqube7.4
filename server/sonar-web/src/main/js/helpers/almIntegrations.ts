/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isLoggedIn } from './users';
import { CurrentUser, AlmOrganization } from '../app/types';

export function hasAdvancedALMIntegration(user: CurrentUser) {
  return (
    isLoggedIn(user) && (isBitbucket(user.externalProvider) || isGithub(user.externalProvider))
  );
}

export function isBitbucket(almKey?: string) {
  return almKey && almKey.startsWith('bitbucket');
}

export function isGithub(almKey?: string) {
  return almKey === 'github';
}

export function isVSTS(almKey?: string) {
  return almKey === 'microsoft';
}

export function isPersonal(organization?: AlmOrganization) {
  return Boolean(organization && organization.personal);
}

export function sanitizeAlmId(almKey?: string) {
  if (isBitbucket(almKey)) {
    return 'bitbucket';
  }
  return almKey;
}
