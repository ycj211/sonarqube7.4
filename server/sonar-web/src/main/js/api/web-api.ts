/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON } from '../helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';

export interface Changelog {
  description: string;
  version: string;
}

export interface Param {
  defaultValue?: string;
  deprecatedKey?: string;
  deprecatedKeySince?: string;
  deprecatedSince?: string;
  description: string;
  exampleValue?: string;
  internal: boolean;
  key: string;
  maximumLength?: number;
  maximumValue?: number;
  maxValuesAllowed?: number;
  minimumLength?: number;
  minimumValue?: number;
  possibleValues?: string[];
  required: boolean;
  since?: string;
}

export interface Action {
  key: string;
  changelog: Changelog[];
  description: string;
  deprecatedSince?: string;
  hasResponseExample: boolean;
  internal: boolean;
  params?: Param[];
  post: boolean;
  since?: string;
}

export interface Domain {
  actions: Action[];
  description: string;
  deprecated: boolean;
  internal: boolean;
  path: string;
  since?: string;
}

export interface Example {
  example: string;
  format: string;
}

export function fetchWebApi(showInternal = true): Promise<Domain[]> {
  return getJSON('/api/webservices/list', { include_internals: showInternal })
    .then(r =>
      r.webServices.map((domain: any) => {
        const deprecated = !domain.actions.find((action: any) => !action.deprecatedSince);
        const internal = !domain.actions.find((action: any) => !action.internal);
        return { ...domain, deprecated, internal };
      })
    )
    .catch(throwGlobalError);
}

export function fetchResponseExample(domain: string, action: string): Promise<Example> {
  return getJSON('/api/webservices/response_example', { controller: domain, action }).catch(
    throwGlobalError
  );
}
