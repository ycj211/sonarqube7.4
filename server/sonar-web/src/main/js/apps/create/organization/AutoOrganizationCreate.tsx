/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import * as classNames from 'classnames';
import { FormattedMessage } from 'react-intl';
import AutoOrganizationBind from './AutoOrganizationBind';
import RemoteOrganizationChoose from './RemoteOrganizationChoose';
import OrganizationDetailsForm from './OrganizationDetailsForm';
import { Query } from './utils';
import RadioToggle from '../../../components/controls/RadioToggle';
import { DeleteButton } from '../../../components/ui/buttons';
import {
  AlmApplication,
  AlmOrganization,
  AlmUnboundApplication,
  Organization,
  OrganizationBase
} from '../../../app/types';
import { bindAlmOrganization } from '../../../api/alm-integration';
import { sanitizeAlmId } from '../../../helpers/almIntegrations';
import { translate } from '../../../helpers/l10n';
import { getBaseUrl } from '../../../helpers/urls';

export enum Filters {
  Bind = 'bind',
  Create = 'create'
}

interface Props {
  almApplication: AlmApplication;
  almInstallId?: string;
  almOrganization?: AlmOrganization;
  almUnboundApplications: AlmUnboundApplication[];
  boundOrganization?: OrganizationBase;
  className?: string;
  createOrganization: (
    organization: OrganizationBase & { installationId?: string }
  ) => Promise<Organization>;
  onOrgCreated: (organization: string, justCreated?: boolean) => void;
  unboundOrganizations: Organization[];
  updateUrlQuery: (query: Partial<Query>) => void;
}

interface State {
  filter?: Filters;
}

export default class AutoOrganizationCreate extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      filter: props.unboundOrganizations.length === 0 ? Filters.Create : undefined
    };
  }

  handleBindOrganization = (organization: string) => {
    if (this.props.almInstallId) {
      return bindAlmOrganization({
        organization,
        installationId: this.props.almInstallId
      }).then(() => this.props.onOrgCreated(organization, false));
    }
    return Promise.reject();
  };

  handleCancelImport = () => {
    this.props.updateUrlQuery({ almInstallId: undefined, almKey: undefined });
  };

  handleCreateOrganization = (organization: Required<OrganizationBase>) => {
    return this.props
      .createOrganization({
        avatar: organization.avatar,
        description: organization.description,
        installationId: this.props.almInstallId,
        key: organization.key,
        name: organization.name || organization.key,
        url: organization.url
      })
      .then(({ key }) => this.props.onOrgCreated(key));
  };

  handleOptionChange = (filter: Filters) => {
    this.setState({ filter });
  };

  renderContent = (almOrganization: AlmOrganization) => {
    const { almApplication, unboundOrganizations } = this.props;

    const { filter } = this.state;
    const hasUnboundOrgs = unboundOrganizations.length > 0;
    return (
      <div className="boxed-group-inner">
        <div className="huge-spacer-bottom">
          <p className="display-flex-center big-spacer-bottom">
            <FormattedMessage
              defaultMessage={translate('onboarding.import_organization_x')}
              id="onboarding.import_organization_x"
              values={{
                avatar: (
                  <img
                    alt={almApplication.name}
                    className="little-spacer-left"
                    src={`${getBaseUrl()}/images/sonarcloud/${sanitizeAlmId(
                      almApplication.key
                    )}.svg`}
                    width={16}
                  />
                ),
                name: <strong>{almOrganization.name}</strong>
              }}
            />
            <DeleteButton className="little-spacer-left" onClick={this.handleCancelImport} />
          </p>

          {hasUnboundOrgs && (
            <RadioToggle
              name="filter"
              onCheck={this.handleOptionChange}
              options={[
                {
                  label: translate('onboarding.import_organization.create_new'),
                  value: Filters.Create
                },
                {
                  label: translate('onboarding.import_organization.bind_existing'),
                  value: Filters.Bind
                }
              ]}
              value={filter}
            />
          )}
        </div>

        {filter === Filters.Create && (
          <OrganizationDetailsForm
            onContinue={this.handleCreateOrganization}
            organization={almOrganization}
            submitText={translate('onboarding.import_organization.import')}
          />
        )}
        {filter === Filters.Bind && (
          <AutoOrganizationBind
            onBindOrganization={this.handleBindOrganization}
            unboundOrganizations={unboundOrganizations}
          />
        )}
      </div>
    );
  };

  render() {
    const { almInstallId, almOrganization, boundOrganization, className } = this.props;

    return (
      <div className={classNames('boxed-group', className)}>
        <div className="boxed-group-header">
          <h2>{translate('onboarding.import_organization.import_org_details')}</h2>
        </div>

        {almInstallId && almOrganization && !boundOrganization ? (
          this.renderContent(almOrganization)
        ) : (
          <RemoteOrganizationChoose
            almApplication={this.props.almApplication}
            almInstallId={almInstallId}
            almOrganization={almOrganization}
            almUnboundApplications={this.props.almUnboundApplications}
            boundOrganization={boundOrganization}
          />
        )}
      </div>
    );
  }
}
