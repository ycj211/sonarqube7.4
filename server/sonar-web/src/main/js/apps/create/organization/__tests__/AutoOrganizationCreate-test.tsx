/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import AutoOrganizationCreate from '../AutoOrganizationCreate';
import { waitAndUpdate, click } from '../../../../helpers/testUtils';
import { bindAlmOrganization } from '../../../../api/alm-integration';

jest.mock('../../../../api/alm-integration', () => ({
  bindAlmOrganization: jest.fn().mockResolvedValue({})
}));

const organization = {
  avatar: 'http://example.com/avatar',
  description: 'description-foo',
  key: 'key-foo',
  name: 'name-foo',
  url: 'http://example.com/foo'
};

it('should render with import org button', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('should render prefilled and create org', async () => {
  const createOrganization = jest.fn().mockResolvedValue({ key: 'foo' });
  const onOrgCreated = jest.fn();
  const wrapper = shallowRender({
    almInstallId: 'id-foo',
    almOrganization: { ...organization, personal: false },
    createOrganization,
    onOrgCreated
  });

  expect(wrapper).toMatchSnapshot();

  wrapper.find('OrganizationDetailsForm').prop<Function>('onContinue')(organization);
  await waitAndUpdate(wrapper);

  expect(createOrganization).toBeCalledWith({ ...organization, installationId: 'id-foo' });
  expect(onOrgCreated).toBeCalledWith('foo');
});

it('should allow to cancel org import', () => {
  const updateUrlQuery = jest.fn().mockResolvedValue({ key: 'foo' });
  const wrapper = shallowRender({
    almInstallId: 'id-foo',
    almOrganization: { ...organization, personal: false },
    updateUrlQuery
  });

  click(wrapper.find('DeleteButton'));
  expect(updateUrlQuery).toBeCalledWith({ almInstallId: undefined, almKey: undefined });
});

it('should display choice between import or creation', () => {
  const wrapper = shallowRender({
    almInstallId: 'id-foo',
    almOrganization: { ...organization, personal: false },
    unboundOrganizations: [organization]
  });
  expect(wrapper).toMatchSnapshot();

  wrapper.find('RadioToggle').prop<Function>('onCheck')('create');
  wrapper.update();
  expect(wrapper.find('OrganizationDetailsForm').exists()).toBe(true);

  wrapper.find('RadioToggle').prop<Function>('onCheck')('bind');
  wrapper.update();
  expect(wrapper.find('AutoOrganizationBind').exists()).toBe(true);
});

it('should bind existing organization', async () => {
  const onOrgCreated = jest.fn();
  const wrapper = shallowRender({
    almInstallId: 'id-foo',
    almOrganization: { ...organization, personal: false },
    onOrgCreated,
    unboundOrganizations: [organization]
  });

  wrapper.find('RadioToggle').prop<Function>('onCheck')('bind');
  wrapper.update();
  wrapper.find('AutoOrganizationBind').prop<Function>('onBindOrganization')('foo');
  expect(bindAlmOrganization as jest.Mock<any>).toHaveBeenCalledWith({
    installationId: 'id-foo',
    organization: 'foo'
  });
  await waitAndUpdate(wrapper);
  expect(onOrgCreated).toHaveBeenCalledWith('foo', false);
});

function shallowRender(props: Partial<AutoOrganizationCreate['props']> = {}) {
  return shallow(
    <AutoOrganizationCreate
      almApplication={{
        backgroundColor: '#0052CC',
        iconPath: '"/static/authbitbucket/bitbucket.svg"',
        installationUrl: 'https://bitbucket.org/install/app',
        key: 'bitbucket',
        name: 'BitBucket'
      }}
      almUnboundApplications={[]}
      createOrganization={jest.fn()}
      onOrgCreated={jest.fn()}
      unboundOrganizations={[]}
      updateUrlQuery={jest.fn()}
      {...props}
    />
  );
}
