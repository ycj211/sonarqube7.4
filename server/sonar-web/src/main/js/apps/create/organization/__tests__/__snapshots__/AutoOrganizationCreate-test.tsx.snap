// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display choice between import or creation 1`] = `
<div
  className="boxed-group"
>
  <div
    className="boxed-group-header"
  >
    <h2>
      onboarding.import_organization.import_org_details
    </h2>
  </div>
  <div
    className="boxed-group-inner"
  >
    <div
      className="huge-spacer-bottom"
    >
      <p
        className="display-flex-center big-spacer-bottom"
      >
        <FormattedMessage
          defaultMessage="onboarding.import_organization_x"
          id="onboarding.import_organization_x"
          values={
            Object {
              "avatar": <img
                alt="BitBucket"
                className="little-spacer-left"
                src="/images/sonarcloud/bitbucket.svg"
                width={16}
              />,
              "name": <strong>
                name-foo
              </strong>,
            }
          }
        />
        <DeleteButton
          className="little-spacer-left"
          onClick={[Function]}
        />
      </p>
      <RadioToggle
        disabled={false}
        name="filter"
        onCheck={[Function]}
        options={
          Array [
            Object {
              "label": "onboarding.import_organization.create_new",
              "value": "create",
            },
            Object {
              "label": "onboarding.import_organization.bind_existing",
              "value": "bind",
            },
          ]
        }
        value={null}
      />
    </div>
  </div>
</div>
`;

exports[`should render prefilled and create org 1`] = `
<div
  className="boxed-group"
>
  <div
    className="boxed-group-header"
  >
    <h2>
      onboarding.import_organization.import_org_details
    </h2>
  </div>
  <div
    className="boxed-group-inner"
  >
    <div
      className="huge-spacer-bottom"
    >
      <p
        className="display-flex-center big-spacer-bottom"
      >
        <FormattedMessage
          defaultMessage="onboarding.import_organization_x"
          id="onboarding.import_organization_x"
          values={
            Object {
              "avatar": <img
                alt="BitBucket"
                className="little-spacer-left"
                src="/images/sonarcloud/bitbucket.svg"
                width={16}
              />,
              "name": <strong>
                name-foo
              </strong>,
            }
          }
        />
        <DeleteButton
          className="little-spacer-left"
          onClick={[Function]}
        />
      </p>
    </div>
    <OrganizationDetailsForm
      onContinue={[Function]}
      organization={
        Object {
          "avatar": "http://example.com/avatar",
          "description": "description-foo",
          "key": "key-foo",
          "name": "name-foo",
          "personal": false,
          "url": "http://example.com/foo",
        }
      }
      submitText="onboarding.import_organization.import"
    />
  </div>
</div>
`;

exports[`should render with import org button 1`] = `
<div
  className="boxed-group"
>
  <div
    className="boxed-group-header"
  >
    <h2>
      onboarding.import_organization.import_org_details
    </h2>
  </div>
  <withRouter(RemoteOrganizationChoose)
    almApplication={
      Object {
        "backgroundColor": "#0052CC",
        "iconPath": "\\"/static/authbitbucket/bitbucket.svg\\"",
        "installationUrl": "https://bitbucket.org/install/app",
        "key": "bitbucket",
        "name": "BitBucket",
      }
    }
    almUnboundApplications={Array []}
  />
</div>
`;
